server/src/main/java/org/opensearch/gateway/PrimaryShardBatchAllocator.java

```diff
@@ -0,0 +1,170 @@
+        // identify ineligible shards
+        for (ShardRouting shard : shards) {
+            AllocateUnassignedDecision decision = getInEligibleShardDecision(shard, allocation);
+            if (decision != null) {
+                inEligibleShards.add(shard);
+                shardAllocationDecisions.put(shard, decision);
+            } else {
+                eligibleShards.add(shard);
+            }
+        }
+        // Do not call fetchData if there are no eligible shards
+        if (eligibleShards.isEmpty()) {
+            return shardAllocationDecisions;
+        }
+        // only fetch data for eligible shards
+        final FetchResult<NodeGatewayStartedShardsBatch> shardsState = fetchData(
+            eligibleShards,
+            inEligibleShards,
+            allocation
+        );
+        // Note : shardsState contain the Data, there key is DiscoveryNode but value is Map<ShardId,
+        // NodeGatewayStartedShardsBatch> so to get one shard level data (from all the nodes), we'll traverse the map
+        // and construct the nodeShardState along the way before making any allocation decision. As metadata for a
+        // particular shard is needed from all the discovery nodes.
+
+        // process the received data
+        for (ShardRouting unassignedShard : eligibleShards) {
+            if (shardsState.hasData() == false) {
+                // if fetching is not done, add that no decision in the resultant map
+                allocation.setHasPendingAsyncFetch();
+                List<NodeAllocationResult> nodeDecisions = null;
+                if (explain) {
+                    nodeDecisions = buildDecisionsForAllNodes(unassignedShard, allocation);
+                }
+                shardAllocationDecisions.put(
+                    unassignedShard,
+                    AllocateUnassignedDecision.no(AllocationStatus.FETCHING_SHARD_DATA, nodeDecisions)
+                );
+            } else {
+
+                List<NodeShardState> nodeShardStates = adaptToNodeShardStates(unassignedShard, shardsState);
+                // get allocation decision for this shard
+                shardAllocationDecisions.put(unassignedShard, getAllocationDecision(unassignedShard, allocation, nodeShardStates, logger));
+            }
+        }
+        return shardAllocationDecisions;
+    }
+
+    /**
+     * shardsState contain the Data, there key is DiscoveryNode but value is Map<ShardId,
+     * NodeGatewayStartedShardsBatch> so to get one shard level data (from all the nodes), we'll traverse the map
+     * and construct the nodeShardState along the way before making any allocation decision. As metadata for a
+     * particular shard is needed from all the discovery nodes.
+     *
+     * @param unassignedShard unassigned shard
+     * @param shardsState fetch data result for the whole batch
+     * @return shard state returned from each node
+     */
+    private static List<NodeShardState> adaptToNodeShardStates(
```

Comment:
We just have to adapt to the old `NodeGatewayStartedShards` here and we can get rid of `NodeShardState` altogether.

Reply:
Done in previous commit.